mdloader_common.c

```diff
@@ -352,6 +352,65 @@ int test_mcu(char silent)
     return 1;
 }
 
+#define DSU 0x41002000
+#define DSU_STATUSB (DSU + 2)
+
+#define NVMCTRL 0x41004000
+#define NVMCTRL_CTRLA (NVMCTRL)
+#define NVMCTRL_CTRLB (NVMCTRL + 4)
+#define NVMCTRL_ADDR (NVMCTRL + 0x14)
+
+#define NVMCTRL_CTRLB_CMDEX_KEY 0xA500
+#define NVMCTRL_CTRLB_CMD_WQW 0x4
+#define NVMCTRL_CTRLB_CMD_PBC 0x15
+#define NVMCTRL_CTRLB_CMD_EP 0x0
+
+#define NVMCTRL_USER 0x00804000
+
+int write_user_row(uint32_t *data)
+{
+    uint16_t status = read_half_word(DSU_STATUSB);
+    printf("dsu statusb: %08x\n", status);
+    write_half_word(DSU_STATUSB, status);
+
+    uint32_t cfg = read_word(NVMCTRL_CTRLA); /* clear nvm interrupt status bits */
+    printf("nvm config: %08x\n", cfg);
+    cfg &= ~(0xf0);
+    write_word(NVMCTRL_CTRLA, cfg);
+
+    write_word(NVMCTRL_ADDR, NVMCTRL_USER); /* set user row address */
+    write_half_word(NVMCTRL_CTRLB, (NVMCTRL_CTRLB_CMDEX_KEY|NVMCTRL_CTRLB_CMD_EP)); /* erase page */
+    slp(100);
+    write_half_word(NVMCTRL_CTRLB, (NVMCTRL_CTRLB_CMDEX_KEY|NVMCTRL_CTRLB_CMD_PBC)); /* erase write buffer */
+    slp(100);
+    for(int i = 0; i < 4; i++) {
+        write_word(NVMCTRL_USER + i * 4, data[i]); /* write in the write buffer */
+    }
+    write_word(NVMCTRL_ADDR, NVMCTRL_USER); /* set user row address */
+    write_half_word(NVMCTRL_CTRLB, (NVMCTRL_CTRLB_CMDEX_KEY|NVMCTRL_CTRLB_CMD_WQW)); /* program quad word (128bits) */
+    slp(100);
+    return 0;
+}
+
+int read_user_row(void)
+{
+    uint32_t user_row[4];
+    printf("user row: ");
+    for (int i = 0; i < 4; i++) {
+        user_row[i] = read_word(0x804000 + i * 4);
+        printf("0x%08x ", user_row[i]);
+    }
+    printf("\n");
+    if ((user_row[1] & 0x7f) != 0x12) {
+        printf("SmartEEPROM not configured, proceed\n");
+        user_row[1] |= (0x12);
```

Comment:
Maybe this should be:
```suggestion
        user_row[1] = (user_row[1] & ~0x7fUL) | 0x12;
```
so that it won't produce anything unexpected if the SmartEEPROM was already configured, but with some other settings?